### ISSUE TITLE
[FR] `ltproperties`: Retrievieng the reference content with "no manipulation"

### ISSUE THREAD

**Issue by gusbrs** [latex3/latex2e#1194]
## Brief outline of the enhancement

`ltproperties` offers `\property_ref:nn` and friends as expandable commands to retrieve a reference's content. However, as far as I can tell at least, there's no (reasonable) way to retrieve that content with "no manipulation", which may be a problem if the content of the reference is itself expandable.

`zref` does offer something of the sort by documenting `\zref@extract` to be expandable in exactly two expansion steps.

It would be interesting if `ltproperties` could offer some way to retrieve the contents of the reference exactly as they are stored in the .aux file.

## Minimal example showing the current behaviour

```latex
\RequirePackage{latexbug}
\documentclass{article}

\newcommand{\foo}{bar}
\ExplSyntaxOn
\property_new:nnnn { myprop } { now } { ?? } { \token_to_str:N \foo }
\ExplSyntaxOff

\begin{document}

\RecordProperties{mylabel}{myprop}
% In the aux: \new@label@record{mylabel}{{myprop}{\foo}}

\RefProperty{mylabel}{myprop} % = "bar" of course

\ExplSyntaxOn
\tl_set:Ne \l_tmpa_tl { \property_ref:nn { mylabel } { myprop } }
\tl_show:N \l_tmpa_tl % > \l_tmpa_tl=bar.  Not `\foo`.
\ExplSyntaxOff

\end{document}
```
I tried to expand `\property_ref:nn` step by step, but in the fourth/fifth step I was still digging, at which point I doubted the exact number of steps could be relied upon, even if I eventually did reach `\foo`.

## Minimal example showing the desired new behaviour

Not a MWE, but as mentioned, it would be nice if it was possible to retrieve the reference content with no manipulation.

**Comment by FrankMittelbach** [comment]
@gusbrs what I'm missing above is a use case for your request, i.e., under what circumstances would you need the data a stored in the aux instead of its expanded form?

**Comment by gusbrs** [comment]
> Well it would be certainly possible to define a command which stores the reference in a tl_var. But I'm not quite sure about the use case.
> 
> Why don't you define `\foo` as protected command if you want it to survive expansion? Or write also a `\noexpand`: `{ \token_to_str:N \noexpand \token_to_str:N \foo }`?

I did build a simple example. But the point really is I'm not necessarily in control of the content of reference. It is not uncommon for packages and classes do add code in a reference. I'd suppose usually the relevant macros are protected, but we cannot assume that in principle. In which case, a way to control that expansion may be needed.

So, assuming that ignorance, I'm suggesting for a way to retrieve the content of the reference without having to worry about what is in there. A way to safely retrieve the content of a reference assuming we don't know or control its content.

> what I'm missing above is a use case for your request, i.e., under what circumstances would you need the data a stored in the aux instead of its expanded form?

Same point as the above.

True, granted it is an edge case. But is it really non existent?

I forgot to mention in the OP, but `refcount` also documents its extraction commands to be expandable in exactly two expansion steps. At least someone else thought there is a need for that, so that we can retrieve those contents unmodified.

**Comment by u-fischer** [comment]
> In the MWE, of course... But not in principle, and that's my only point.

Yes, I understand. But this is why I asked for a use case: from your MWE it is unclear if you have a real problem and if a get-command would actually solve that.

**Comment by gusbrs** [comment]
> Yes, I understand. But this is why I asked for a use case: from your MWE it is unclear if you have a real problem and if a get-command would actually solve that.

The use case is the one I described: I don't want to have to know what's in there to retrieve the reference. Of course, I understand what you are asking: "show me a relevant class or package which breaks things"... Well, if you really think there's nothing to worry about, feel free to close the issue. I don't really have more arguments to add than the above.

**Comment by gusbrs** [comment]
Thank you all for taking the time to consider the request.

**Comment by gusbrs** [comment]
I had closed this myself, and understand your arguments, but I was working with some related stuff today and came up with a less contrived MWE which may serve as better motivation to the original request:

```tex
\RequirePackage{latexbug}
\documentclass{article}

\renewcommand*{\thepage}{\textbf{\arabic{page}}}

\begin{document}

\section{Section}
\RecordProperties{mylabel}{page}

\RefProperty{mylabel}{page}

\ExplSyntaxOn
\tl_set:Ne \l_tmpa_tl { \property_ref:nn { mylabel } { page } }
% \tl_show:N \l_tmpa_tl
\ExplSyntaxOff

\end{document}
```
This fails with:

```
./document.tex:14: Undefined control sequence.
\reserved@a ->\@nil 
                    
l.14 ... { \property_ref:nn { mylabel } { page } }
                                                  
The control sequence at the end of the top line
of your error message was never \def'ed. If you have
misspelled it (e.g., `\hobx'), type `I' and the correct
spelling (e.g., `I\hbox'). Otherwise just continue,
and I'll forget about whatever was undefined.
```
[document.log](https://github.com/user-attachments/files/17941342/document.log)

The reason is, of course, that the robust, but not protected `\textbf` is part of the property. And we are trying to expand it with `\tl_set:Ne`. But, considering `ltproperties` does use `\protected@write` to write the labels and this content is technically allowed, wouldn't it make sense to allow for some expansion control in retrieving the property? Specially considering `ltproperties` is particularly suited to "store data" rather than just "typeset a `\ref`".

Of course, I know (now... ;-) I could use `protected@edef` instead of `\tl_set:Ne` there. But the point of the original request still stands: why should we have to worry about the "possible content of the property" simply to retrieve it?

**Comment by u-fischer** [comment]
> `\renewcommand*{\thepage}{\textbf{\arabic{page}}}`

Well you shouldn't do that. With hyperref it works as it takes care of \textbf when building PDF strings and page labels, but for example makeindex will reject index entries with bold page numbers as it can't sort them. You will also get bold page numbers in place where you probably do not want them like with a `\pageref`. 

> why should we have to worry about the "possible content of the property" simply to retrieve it?

You have to worry about the possible content as no property can be used in all places. E.g. if you are in a place where a number is expected you can use the abspage property but not the page or the dest property. 

Even if there is a command to retrieve the property and store it in a macro, you still would have to know something about the content of this macro if you want to process it further. You still would have to know if it will break if used with an \edef or if is can be used as number.

So the question still stands: What is the use case that you need to store it unchanged in a variable? Why is using `\protected@edef` or `\tl_set:Ne \l_tmpa_tl { \text_expand:n{\property_ref:nn { mylabel } { page }} }` not enough?

**Comment by mbertucci47** [comment]
Just a few thoughts on the issue, happy to be contradicted.

I ran into this issue while playing around with the zref-clever code to see how difficult it would be to implement with ltproperties. The answer is, not difficult at all if there was a way to do what's requested in this FR (I'm sure @gusbrs knows this and ran into it while trying to do the same thing with his package). Basically the package doesn't want to break a user's bad input (something like `\def\thesection{\textit{\arabic{section}}}`) but needs access to the property's content in an expansion setting (@gusbrs please correct me if I'm wrong). This means `\protected@edef` won't work. Using `\text_expand:n` as suggested does work, but it is incredibly slow, many orders of magnitude slower than zref-clever's current approach of two expansions on `\zref@extract`.

Since ltproperties allows `\RecordProperties` to write anything to the aux that survives `\protected@write`, I think it makes sense to allow retrieval of the content "as-is", since the initial recording did not require the content to be fully expandable. Of course this is possible with `\protected@edef` as mentioned but a version that works in an expansion context would be nice too.

**Comment by u-fischer** [comment]
@mbertucci47 @gusbrs we have added `\property_item:nn` which will return the value in a `\exp_not:n`.

**Comment by mbertucci47** [comment]
@u-fischer Would it also make sense to add `\property_item:nnn|een` to mirror `\property_ref:nnn|een` ? Looking at the zref-clever code, the author actually uses the discussed facts about number of expansions for `\zref@extractdefault` so I think having the "with default" version would be helpful too.
